### PR TITLE
[FIXED JENKINS-32326] Support no_proxy environment variable

### PR DESCRIPTION
**Pull request by etiennebec**
https://issues.jenkins-ci.org/browse/JENKINS-32326

### CODE REVIEW

File: src/main/java/hudson/remoting/Util.java [hunk 1]
```diff
@@ -108,6 +108,55 @@ static String indent(String s) {
     }
 
     /**
+     * Check if given URL is in the exclusion list defined by the no_proxy environment variable.
+     * On most *NIX system wildcards are not supported but if one top domain is added, all related subdomains will also
```
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Util.java]: does not help much with redirects, etc.

**Comment by oleg-nenashev** [comment]
In Jenkins project we commonly use System properties. IMHO it would be useful to support both options. System property could have a higher priority in this case. On the other hand, such approach may lead to security concerns

File: src/main/java/hudson/remoting/Util.java (hunk 1 above)
Comment (etiennebec) [inline src/main/java/hudson/remoting/Util.java]: You mean HTTP 301/302 redirection? The goal of this function is **only** to determine whether the slave needs to use a proxy or not to reach its master. The actual selection of proxy address, redirection/connection to the proxy will be handled by other parts: `Util.openURLConnection()` relies on `java.net.URL.openConnection()` to do so, and the Engine does it manually: https://github.com/jenkinsci/remoting/blob/remoting-2.59/src/main/java/hudson/remoting/Engine.java#L411

I tried to mimic as much as possible the behaviour of well known tools as curl or wget which supports the `no_proxy` environment variable.

**Comment by etiennebec** [comment]
I would also prefer to only rely on system properties. But previous PR #27 introduced the use of `http_proxy` environment variable for no reason without supporting `no_proxy` exception list (equivalent of `http.nonProxyHosts` but less powerful).

So since #27 was merged, **every** remoting release includes a bug which prevent a slave to connect to its master when `http_proxy` was set. Because even if the system properties have precedence over this variable (see `Util.openURLConnection()` and `Util.getResolvedHttpProxyAddress()`), `http_proxy` was used as a fallback when they were not defined.

To sum up if you must use a proxy to access internet and has configured `no_proxy` so apt/curl/wget for instance could download files, your slave cannot currently contact its master which is on the same network.

Please have a look to the last comments of [JENKINS-28289](https://issues.jenkins-ci.org/browse/JENKINS-28289?focusedCommentId=258469&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-258469) and [JENKINS-32326](https://issues.jenkins-ci.org/browse/JENKINS-32326) for further details.

File: src/main/java/hudson/remoting/Util.java (hunk 1 above)
Comment (oleg-nenashev) [inline src/main/java/hudson/remoting/Util.java]: Agreed, it should work

**Comment by oleg-nenashev** [comment]
LGTM 👍 
Needs some streamlining for System Properties, but it's a follow-up task IMHO
